Sites Page: Fix site picker choking when path does not include registered route

mariovalney opened the pull request:
**Problem:**

Issue #11543 

1. When we are in `/sites` the site picker works.
2. But in `/sites/mywebsite.com` it change the current site, but that's all because `/sites/mywebsite.com` is equal to `/sites/anothersite.com`.

**Solution:**

We can fix the routing by:

- Redirecting `/sites/mywebsite.com` to `/sites` (like `/sites/invalid.com`) or to `/stats/insights/mywebsite.com` like a shortcut.
- Considering `/sites/mywebsite.com` like `/sites` and onClick will go to `/stats/insights/selectedsite.com` (that's the proposed solution in this initial PR).
- Consider it a 404

**Another Problem:**

Site picker on `/settings/mywebsite.com` fails too.

**Solution:**

- Redirect `/settings/mywebsite.com` to site settings (`/settings/general/mywebsite.com`)
- Considering `/settings/mywebsite.com` like `/settings/general` and onClick will go to `/settings/general/selectedsite.com` (that's the proposed solution in this initial PR).
- Consider to be 404

**EDIT:**

1 - Just for clarifying: each item of list is a option to fix the problem not a step.
2 - Rewrited list of solutions to be more clear

Commit:
Fix: consider /sites/site.com and /settings/site.com as section without site.com

mariovalney commented:
This commit (5fa59a2) fix the path sended to `addSiteFragment` to consider `/sites/site.com` as `/sites` (sending user to `/stats/insights`) and `/settings/site.com` as `/settings` (sending user to `settings/general`).

dmsnell commented:
Thanks for proposing this fix! If we're already at `/stats/insights` will the error reappear? I'm sorry because I'm having a little trouble understanding the problem and solution. It sounds like because our sections stay the same the app isn't re-rendering once we select a new site. Is that wrong?

mariovalney commented:
Hi! Will check this now but I guess it'll not reappear.

[ A few minutes after... ]

Nope: it's ok on site picker at `/stats/insights` after navigating from `/sites/mysite.com`.

alisterscott commented:
Thanks for the fix - I tested this and it works for me in that it shows the site picker and selecting a site now works.

It does look weird when there is just one site to have to select one:

<img width="395" alt="screen shot 2018-06-25 at 11 26 07 am" src="https://user-images.githubusercontent.com/128826/41826269-89b46476-786b-11e8-8375-88e03019a4a7.png">

<img width="405" alt="screen shot 2018-06-25 at 11 25 32 am" src="https://user-images.githubusercontent.com/128826/41826271-8f3b4482-786b-11e8-9bcd-a74eb5d18c7d.png">

Give that, I think this would be the best solution:

If navigating to https://wordpress.com/sites/sitethatexists.com **where the site exists** - then automatically show https://wordpress.com/stats/insights/sitethatexists.com - so a redirect in essense

If navigating to https://wordpress.com/sites/sitethatdoesnotexist.com **where the site does not exist** - then show https://wordpress.com/stats if the user has multiple sites, or automatically show https://wordpress.com/stats/insights/sitethatexists.com if the user only has a single site (to avoid the situation above)

Same behaviour for settings, but show settings/general

Does that make sense? Is this workable?

mariovalney commented:
Hi! Yes it makes sense (the fewer clicks the better to user).
Just to be sure: do you think is better redirect just when user has only one site, didn't you?

So... maybe we should consider this fix and than start to discuss/implement this behaviour at any point of site picker?

alisterscott commented:
> Just to be sure: do you say is better redirect just when user has only one site?

Yes, I think this makes most sense - can this PR continue in this direction?

mariovalney commented:
Yes, of course. 

My only concern is to keep consistency. So I'll take a better look to check if there is another point like settings/sites.

mariovalney commented:
Hey @alisterscott, could you help me in a point?
When developing I'm using my own account. What is the best way to manage these special cases, for example, simulate just one site?

_I had to remove a ununsed site to keep only one to test_ :flushed:

Commit:
Fix: /sites and /settings redirect when there is only one site

mariovalney commented:
Hi again! 
_sorry for the 3 notifications hahaha_

That's done, I guess... 
I intercepted the pathname in the [redirectToPrimary](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/controller.js#L340) function to check it's only `/settings` and added the `general` of it's `/sites` and change to stats.

As we have a IF to check the `siteFragment` is empty (avoiding loops), I added a condition to redirect if we are in that pages. So it will work to `/sites/myonlysite.com` and `/settings/myonlysite.com`.

alisterscott commented:
> What is the best way to manage these special cases, for example, simulate just one site?

It's probably best to just create a separate free account with a single test site, if you use GMail you can use GMail + addressing to use the same GMail account for your second account.

alisterscott commented:
Thanks for the update. I retested this:
1. When no site is specified and the customer has multiple sites it shows the picker which now works 👍 
2. When no site is specified and the customer only has one site it straight away shows the correct page not the picker 👍 
3. When a site is specified and customer has multiple sites, it shows the picker which works - do we expect this to redirect straight to stats/insights, or settings/general like it would for when the customer only has one site❓
4. When a site is specified and the customer only has a single site it straight away shows the correct page not the picker 👍 

Only outstanding question from my testing. Thanks

jsnajdr commented:
Wow, the site selection code got really incomprehensible over the last few years.

This is how site selection originally looked 4 years ago (source: non-public 1253-gh-calypso-pre-oss):

![a12bf848-57c8-11e4-93d9-ca627b9a2528](https://user-images.githubusercontent.com/664258/41964364-0e2e76e4-79fa-11e8-9f00-f24573fd8c2c.png)

There used to be a filterbar with "All Sites", "Public", "Private" and search. That's been replaced with the `SiteSelector` component and many other changes happened that made lot of code obsolete.

I think this PR adds even more confusion to an already obfuscated code. I'd prefer if we did a series of cleanups instead that will greatly improve the code quality:

In `sites.jsx`:
1. Remove the `observe( 'user' )` mixin: the component doesn't access the User object at all. The code that did was removed a long time ago.
2. Codemod the `createReactClass` into an ES6 React class.
3. The component doesn't need the `selectedSite` prop. It's used in the `filterSites` but that's obsolete code. It used to reset the "Public"/"Private" filters in filterbar when a site was selected. The UI is completely different today.
4. The component doesn't need the `sites` prop and doesn't need to pass it down to `SiteSelector`. `SiteSelector` has its own `connect` call that retrieves `sites` from Redux.
5. Can we remove `client/my-sites/sites/package.json` and rename `sites.jsx` to `index.jsx`? Specifying a non-default index file in `package.json` is a convention we no longer use.
6. The `filterSites` method is supposed to filter sites according to some criteria (public, private, upgradeable, Jetpack, ...), but the `SiteSelector` also shows recent sites that don't qualify for the criteria. That's confusing. Can we refactor the component in such a way that `SiteSelector` has prop `groups={false}` (which enables the recent sites list) when a filter is active?
7. The public/private filtering is no longer needed, as well as the `sitesFilter` route param. It was useful when the UI included the filterbar, but now there is no UI that points to the `/sites/private` and `/sites/public` routes.

The `SiteSelector` component also has a neat feature: when passed a `siteBasePath` prop:
```jsx
<SiteSelector siteBasePath="/stats/insights" />
```
then it doesn't need the `onSiteSelect` prop and will handle site selection by navigating to `/stats/insights/:site_slug`. That's all we need, don't we? The `selectSite` action can be removed, as it does the same thing.

In the `createSitesComponent` function in the controller, there is some mysterious code that checks `context.prevPath` and computes `path` and `sourcePath` props for `SitesComponent`. Is that necessary? It seems to me that we could just pass one `path={ sectionify( context.path ) }` prop to `SitesComponent` and be done. Is there any distinction between `path` and `sourcePath`?

After the `sitesFilter` param is removed from the `/sites/:sitesFilter?` route, the original bug could be solved with a simple redirect:
```js
page( '/sites/:site`, context => page.redirect( '/stats/insights/' + context.params.site ) );
```
If the site slug is in the URL, why show a site selection UI? Let's redirect to stats right away.

The issues with `/settings` site selection could be solved similarly: by improved handling of the `/settings/*` routes:
- there should be a redirect from `/settings/site.slug` to `/settings/general/site.slug` if `site.slug` is a valid site slug (number or string with dots)
- this redirect should take care to coexist well with the `/settings/:section` handler that does legacy redirects

@mariovalney If you want to work on these issues, I think they are a good fit for an OSS contributor who is not terribly familiar with the code base. It's a series of small PR that all fix one simple issue at a time.

mariovalney commented:
Hi @jsnajdr ! thanks for your points.
I have to agree with you: it's a bit confused.

So.. to work on that we move this points to a Issue and start to create new PRs?

jsnajdr commented:
> So.. to work on that we move this points to a Issue and start to create new PRs?

Yes, that would be a series of standalone PRs.

However, if you want to just fix the original redirect issue, this is what I think is a minimal path to a correct fix:
1. Remove the support for `/sites/public` and `/sites/private` filters and leave just a handler for `/sites`.
2. Create a `/sites/:site` redirect handler as described above.
3. Fix the `/settings` issue by implementing correct redirects in the Settings controller.

The remaining issues I mentioned are optional, although very very desirable. Cleaning up the site selector would be an awesome contribution 🥇

mariovalney commented:
Well. I guess is better to close that and start from cleaning up indeed.
I created the #25783 issue to help us to organize this.

Maybe can we close #11543 and point it to there?

dmsnell commented:
Wow! What a difficult problem lying underneath. @mariovalney if you are interested in tackling it then we'll do what we can to help out. Sounds like @jsnajdr's recommendation will solve the problem without causing too much deferred maintenance 😄
